src/java.base/share/classes/sun/security/provider/DSAParameterGenerator.java

```diff
@@ -235,15 +235,15 @@ private static BigInteger[] generatePandQ(SecureRandom random, int valueL,
                 /* Step 11.2 */
                 BigInteger W = V[0];
                 for (int i = 1; i < n; i++) {
-                    W = W.add(V[i].multiply(BigInteger.TWO.pow(i * outLen)));
+                    W = W.add(V[i].shiftLeft(i * outLen));
                 }
                 W = W.add((V[n].mod(BigInteger.TWO.pow(b)))
```

Comment:
Did you want to update the "BigInteger.TWO.pow(b)" as well?

Reply:
Actually I think I'll revert the DSA changes; I'm not prepared to measure their effect at this moment.